src/data/nat/cast/defs.lean

```diff
@@ -0,0 +1,156 @@
+/-
+Copyright (c) 2014 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Gabriel Ebner
+-/
+import algebra.group.defs
+import algebra.group.basic
+
+/-!
+# Cast of natural numbers
+
+This file defines the *canonical* homomorphism from the natural numbers into an
+`add_monoid` with a one.  In additive monoids with one, there exists a unique
+such homomorphism and we store it in the `nat_cast : ℕ → R` field.
+
+Preferentially, the homomorphism is written as a coercion.
+
+## Main declarations
+
+* `add_monoid_with_one`: Type class for `nat.cast`.
+* `nat.cast`: Canonical homomorphism `ℕ → R`.
+-/
+
+universes u
+set_option old_structure_cmd true
+
+/-- The numeral `((0+1)+⋯)+1`. -/
+protected def nat.unary_cast {R : Type u} [has_one R] [has_zero R] [has_add R] : ℕ → R
+| 0 := 0
+| (n + 1) := nat.unary_cast n + 1
+
+/--
+An `add_monoid_with_one` is an `add_monoid` with a `1`.
+It also contains data for the unique homomorphism `ℕ → R`.
+-/
+@[protect_proj]
+class add_monoid_with_one (R : Type u) extends add_monoid R, has_one R :=
+(nat_cast : ℕ → R := nat.unary_cast)
```

Comment:
How much breaks if we remove this default and force downstream code to use `.unary` or `.binary` instead if they want this behavior?